Bio/File.py

```diff
@@ -17,16 +17,80 @@
 import collections.abc
 
 from abc import ABC, abstractmethod
+from os import PathLike
+from typing import IO, Iterator, Any, AnyStr, Union, Tuple, Optional
+
+from Bio import StreamModeError
 
 try:
     import sqlite3
 except ImportError:
     # May be missing if Python was compiled from source without its dependencies
     sqlite3 = None  # type: ignore
 
+# https://docs.python.org/3/glossary.html#term-path-like-object
+_PathLikeTypes = (PathLike, str, bytes)
+_IOSource = Union[IO[AnyStr], PathLike, str, bytes]
+_TextIOSource = _IOSource[str]
+
+
+def check_handle_mode(handle: IO, mode: str, fmt: Optional[str] = None) -> None:
+    """
+    Check if handle is opened with mode.
+
+    Raises:
+      StreamModeError if handle is not opened with mode
+    """
+    try:
+        if "w" in mode or "a" in mode or "+" in mode:
+            if "b" in mode:
+                msg = "must be opened for writing in binary mode"
+                handle.write(b"")
+            else:
+                msg = "must be opened for writing in text mode"
+                handle.write("")
+            if "+" in mode:
+                msg = "must be opened for updating"
+                handle.read(0)
+        elif "b" in mode:
+            msg = "must be opened in binary mode"
+            if not isinstance(handle.read(0), bytes):
+                raise TypeError
+        else:
+            msg = "must be opened in text mode"
+            if not isinstance(handle.read(0), str):
+                raise TypeError
+    except (TypeError, OSError):
+        prefix = f"{fmt} files " if fmt else "File "
+        raise StreamModeError(prefix + msg) from None
+
+
+def as_handle_and_flag(
+    handleish: _IOSource, mode: str = "r", *, fmt: Optional[str] = None
+) -> Tuple[IO, bool]:
+    """
+    Open handleish with mode, or return it as-is if it's already a file handle.
+
+    Args:
+      handleish: File handle or path-like object
+      mode: Mode to open handleish
+      fmt: File type name for error message
+
+    Returns:
+      File handle and flag whether the caller should close the handle or not.
+
+    Raises:
+      StreamModeError if handleish is a file handle and is not opened with mode
+    """
+    if isinstance(handleish, _PathLikeTypes):
+        return open(handleish, mode), True
+
+    check_handle_mode(handleish, mode, fmt)
```

Comment:
If the FASTA parser is to accept both text and binary streams in the future, this check could get in the way.

Reply:
Unless we do:
```py
    if "b" in mode:
        handleish = getattr(handleish, "buffer", handleish)
```

Reply:
That probably depends on how we end up making the FASTA (etc) parser cope with bytes. Probably a high level switch between unicode and bytes versions.